DAS-2214 Add NSIDC IceSAT2 Regression Tests

flamingbear opened the pull request:
## Description

This PR adds a new regression test to the repository.

This is a stop-gap to allow NSIDC to test the basic IceSAT2 Collections against Harmony Services (Primarily the Trajectory Subsetter).

This PR adds:
 
- a set of regression tests in the `test/nsidc-icesat2/` directory
- a `shared_utility` directory so that common basic functionality can be shared across tests.  Right now there are a number of basic routines copied into each tests directory, they can be migrated out slowly (or all at once in an IP sprint).
- github workflow infrastructure to test and build the nsidc-icesat2-image 


## Jira Issue ID
[DAS-2214](https://bugs.earthdata.nasa.gov/browse/DAS-2214)

## Local Test Steps

This can be tested in UAT so in order to test locally, you must clone and check out this branch.
 
 Build the image from the `test` directory:
``` shell
make nsidc-icesat2-image
```

Set your UAT environmental variables

``` shell
# UAT
export HARMONY_HOST_URL=https://harmony.uat.earthdata.nasa.gov
export AWS_ACCESS_KEY_ID=<UATAWSKEY>
export AWS_SECRET_ACCESS_KEY=<UATAWSSECRETKEY>
export EDL_USER=<edl username>
export EDL_PASSWORD=<edl password>
```

Run the notebook against UAT

``` shell
./run_notebooks.sh nsidc-icesat2
```

Verify it run successfully

output looks like:
``` shell
Running regression tests
Test suite nsidc-icesat2 starting
running test with ghcr.io/nasa/regression-tests-nsidc-icesat2:latest
Waiting for nsidc-icesat2.
Input Notebook:  nsidc-icesat2/NSIDC-ICESAT2_Regression.ipynb
Output Notebook: /workdir/output/nsidc-icesat2/Results.ipynb
Working directory: nsidc-icesat2
Executing:   0%|          | 0/18 [00:00<?, ?cell/s]Executing notebook with kernel: python3
Executing: 100%|██████████| 18/18 [09:26<00:00, 31.50s/cell]
Test suite nsidc-icesat2 succeeded
Tests completed (passed)
```

and includes `Test suite nsidc-icesat2 succeeded`

## PR Acceptance Checklist
* [X] Acceptance criteria met
* [X] Tests added/updated (if needed) and passing
* [X] Documentation updated (if needed)

asteiker commented on test/nsidc-icesat2/NSIDC-ICESAT2_Regression.ipynb:
````diff
@@ -0,0 +1,433 @@
+    "```sh\n",
+    "conda env create -f ./environment.yaml && conda activate papermill-nsidc-icesat2\n",
+    "```\n",
+    "\n",
````
```suggestion
    "\n",
"To use this environment within a shared Jupyter Hub, see [instructions](https://nasa-openscapes.github.io/earthdata-cloud-cookbook/contributing/workflow.html#create-a-jupyter-kernel-to-run-notebooks) in the NASA Earthdata Cloud Cookbook for how to create a new kernel based on this environment.

asteiker replied:
I tried testing this within the Openscapes 2i2c Jupyter Hub and needed these extra steps to create the kernel. This then worked for me.

flamingbear replied:
This is cool, I haven't run anything in openscapes.

asteiker commented on test/nsidc-icesat2/NSIDC-ICESAT2_Regression.ipynb:
```diff
@@ -0,0 +1,433 @@
+    "                f'reference_files/{test_output.stem}_reference{test_output.suffix}'\n",
+    "            )\n",
+    "\n",
+    "            submit_and_download(harmony_client, test_request, test_output)\n",
```
@flamingbear I got an error here when running locally: 

`OSError: [Errno 18] Invalid cross-device link: '4720896_ATL03_20200427193622_04930702_006_02_subsetted.h5' -> '[/tmp/tmp3bpwuc_y/ATL03_subset_bounding_box.h5](https://openscapes.2i2c.cloud/tmp/tmp3bpwuc_y/ATL03_subset_bounding_box.h5)'`

flamingbear replied:
This is probably because you're not actually local, but when I'm back I'll see if [this stackoverflow](https://stackoverflow.com/a/43967659/66100) answer can fix this for you

flamingbear replied:
So I think this 89cae89 might fix it. It isn't hurting the existing local and docker runs. If you get a chance you can update your branch and try again? 🙇

asteiker replied:
Looks like that did the trick. Notebook is running end to end now in the Openscapes Hub. 🎉

lyonthefrog reviewed (commented):
I ran the test instructions and the tests passed! My comments are mainly a few questions and a couple minor typos. I'll approve once the questions are resolved.

flamingbear commented on test/shared_utils/utilities.py:
You nailed it.  This is a concern and why I waffled on this a bit.  But, there are some basic things that all tests need to do and we probably just need to curate the functions in the shared utils carefully.

lyonthefrog replied:
I'm thinking an issue that might be encountered is using the comparison function for different file formats, since I know I hit that when trying to add that function to the `sdps-egi-tests` repository. But it would be nice to have. Would there be any one "owner" of these utils, or set of rules for what should and should not be added? But I'm likely overthinking this and perhaps the answer is that we trust the folks using this repository to be aware that any change may affect the rest.

flamingbear replied:
Yup, I think custom comparisons should live in their own test directory.  But yes, this is just a be careful spot I think.

flamingbear commented on test/shared_utils/utilities.py:
```diff
@@ -0,0 +1,148 @@
+except Exception:
+    # only used by Trajectory Subsetter tests.
+    # TODO: remove and make Trajectory Subsetter use above
+    from datatree import open_datatree
```
This is not fully dependent on the xarray work, but I didn't want to update the Trajectory Subsetter in this PR.  I mean a ticket would be good, but I'll probably just sneak it in during the IP. If not, it should be handled the next time the trajectory subsetter tests need to change.

flamingbear commented on test/nsidc-icesat2/environment.yaml:
```diff
@@ -0,0 +1,13 @@
+  - pip
+  - pip:
+    - harmony-py==0.4.15
+    - git+https://github.com/pydata/xarray.git@ca2e9d6#egg=xarray
```
Yes, this is a git link to a specific hash from that repository.  Because open_groups isn't in any released versions yet, I need to do this.

lyonthefrog replied:
Neat!

flamingbear commented on README.md:
````diff
@@ -90,12 +90,15 @@ environment before installing from the environment.yml.
    ```
    <new-suite-name>-image
-       docker build -t ghcr.io/nasa/regression-tests-<new-suite-name>:latest -f ./Dockerfile --build-arg notebook=<new-test-notebook-name> --build-arg sub_dir=<new-suite-subdirectory> .
+       docker build -t ghcr.io/nasa/regression-tests-<new-suite-name>:latest -f ./Dockerfile --build-arg notebook=<new-test-notebook-name> --build-arg sub_dir=<new-suite-subdirectory> [--build-arg shared_utils=true] .
````
Hmm, I'm second guessing.  I meant to show that you could include the shared_utils directory by including that optional argument.  But it's not true by default.

lyonthefrog replied:
Got it! I think I'm just used to seeing opt-arg=default when it's written out like this, but this is just an example in the README so I'm sure it's fine. What sets the default value?

flamingbear replied:
It's an optional argument, so there's no default value, the default value is noting, but the pseudo-equivalent would be false.

flamingbear commented on test/Dockerfile:
```diff
@@ -20,9 +22,24 @@ COPY build-netrc.sh notebook-entrypoint.sh ./
+COPY shared_utils ./shared_utils
+RUN if [ "$shared_utils" = "false" ]; then \
+        rm -rf ./shared_utils; \
+    fi
```
It's a long stupid story, but it ends with you can run a command in a conditional 

```
RUN if [ test ]; run a command in the container
```

But you can't COPY in a conditional so I have to copy the dir every time (can't conditionally do this command)
```
COPY shared_utils ./shared_utils 
```
and then conditionally run a command to delete the files.

I'm very open to suggestions but google came up with this.

lyonthefrog replied:
Meh, I searched briefly on my end and I found the same sort of thing. This works for all intents and purposes.

flamingbear commented on test/shared_utils/README.md:
```diff
@@ -0,0 +1,52 @@
+
+The test environment is determined by the environment.yaml in the test directory, but if you are including `shared_utils` you will need to also include harmony-py and either xarray-datatree or a fancy pinned version of xarray
+
+For example the pip requirements in the nsidc_icesat2 environment.file :
```
fdf6c6e

flamingbear commented on CHANGELOG.md:
```diff
@@ -5,11 +5,18 @@ versioning. Rather than a static releases, this repository contains of a number
+- Adds NSIDC ICESat2 Regression test suite.
+
+- Adds `shared_utils` functionality. This directory contains routines that are commonly used in regression tests and limits code duplication. To include the `shared_utils` directory in your docker container, Update the Makefile to add a shared_utils build arg. E.g. `--build-arg shared_utils=true` (see the nsidc-icesat2-image target)
+
```
193644dc907a2a52d15aa7328ae313e76bd5ee1e

owenlittlejohns reviewed (commented):
Looks like you fixed up the pre-commit things that had failed while I was working through this review - nice!

This is nice work, and I like the extraction of things into a shared utility module (although feel your pain about the conditional shenanigans you had to go through in the `Dockerfile`).

I haven't given this the big thumbs up because I'm cautious about the sizes of some of the reference files. Are there tests that would give NSIDC enough confidence but using smaller spatial/temporal regions?

owenlittlejohns commented on test/nsidc-icesat2/reference_files/ATL03_subset_bounding_box_reference.h5:
Rambling stream-of-consciousness ahoy:

One thought Vinny suggested a while back - could the reference files go somewhere like S3? Then the notebook could either do some in-cloud access, or download the reference files when being run, maybe? Problem with in-cloud access: these notebooks then need to run in us-west-2, which will be problematic as we look to lean on running things in GH (truly rambling: CodeDeploy in Harmony AWS account associated with this repository?). Problem with downloading when running: slow at runtime, rather than slow once when compiling a Docker image.

---

Another (lower-tech) thought - what's the size of the source file? ATL03 is indeed pretty beefy, but some of the granules are up in the several GB range, while others are more like 300 MB. I wanted to check the ATL03 granule, but don't have access to view CMR records in the NSIDC_CUAT provider, so couldn't. 

This led me to:

1) If these tests are going to be automated, we need to make sure that whichever EDL user is running them (credentials probably set by secrets in a couple of GH environments) has the requisite permissions to view/access/download data and metadata records.

2) The tests currently only define granules for non-production accounts. Are there production equivalents? Don't we need those? (When I couldn't get at the UAT granule for ATL03, I was going to rip the equivalent from production, but that isn't defined, so I was defeated again!)

Sorry - that's probably about 4 comments rolled into one.

flamingbear replied:
I did forget the authentication that would be required (I had a note at one point: and it was that I would add the test user to NSIDC's ACLs I think)

As for the size, I did go back and forth with Mikala, and I can go back to her, but maybe you should if this is a blocker.

flamingbear replied:
I did ask Mikala for clarification on why these were the smallest reference files she felt comfortable making. And we will see what she says

flamingbear commented:
After talking with @asteiker @owenlittlejohns and @mikala-nsidc.  I will update this PR configuring the repository to use github's LFS to keep these large files separate from the code.

flamingbear commented:
To complete:
- [ ] update repo with GIT LFS and store the reference files there.
- [ ] update pinned version of xarray to latest release 
- [ ] clean up docker requirements related to installing git for the previous pin

flamingbear commented:
@owenlittlejohns This is ready for re-review with questions at the bottom

Things I have done:

- verified that this set of commits added all of the reference files.
	92a42e6 | * DAS-2214: Adds ATL08 shapefile reference file
	123bc10 | * DAS-2214: Adds ATL09 ATL10 shapefile reference files
	80ae48b | * DAS-2214: Adds ATL13 BBox reference file.
	3bd883b | * DAS-2214: Add reference files Subset bounding box ATL10 and ATL12
	23d6557 | * DAS-2214: Adds ATL06 shapefile reference file.
	ab651f3 | * DAS-2214: Adds reference file for ATL04 temporal range tests
	12edb03 | * DAS-2214: Adds regression files for NSIDC BBox Subset tests.

- reverted them all (6b23cd2)
- tracked NSIDC's regression test directory with gitlfs.
   `❯ git lfs track "test/nsidc-icesat2/reference_files/*.h5"`

- added the reference files back to the proper location, and committed them and the .gitattributes files that tells git to use lfs for this directory.


- I am 99% sure that when this is merged new clones will not have to get all of the binary files in their history, and eventually github will clean it up.

- I have updated the actions so that lfs is a new option to the build-all-images matrix.

- I also updated the actions to include the build_arg for shared-utils when I found that it wasn't included in my flamingbear fork.

**Lastly, I deleted my fork and when I recreated it I can't test these actions anymore because of a login failure. Maybe you can fork and test yours.**

In the actions:

You can  examine the differences between NSIDC image and any other image and you will see the lfs being used to check out the repository.

```
Checkout regression tests repository for build-images (all other images)
Run actions/checkout@v3
Syncing repository: flamingbear/harmony-regression-tests
Getting Git version info
Temporarily overriding HOME='/home/runner/work/_temp/9f3bf3d5-b39a-4e61-9403-3ad7f60a95d1' before making global git config changes
Checkout regression test repository for build-images (nsidc-icesat2)
Adding repository directory to the temporary git global config as a safe directory
/usr/bin/git config --global --add safe.directory /home/runner/work/harmony-regression-tests/harmony-regression-tests
Deleting the contents of '/home/runner/work/harmony-regression-tests/harmony-regression-tests'
Initializing the repository
Disabling automatic garbage collection
Setting up auth
Fetching the repository
Determining the checkout info
Checking out the ref
/usr/bin/git log -1 --format='%H'


Checkout regression test repository for build-images (nsidc-icesat2)
Run actions/checkout@v3
Syncing repository: flamingbear/harmony-regression-tests
Getting Git version info
Temporarily overriding HOME='/home/runner/work/_temp/7161e340-c13e-4205-a677-e1b79c3ca60b' before making global git config changes
Adding repository directory to the temporary git global config as a safe directory
/usr/bin/git config --global --add safe.directory /home/runner/work/harmony-regression-tests/harmony-regression-tests
Deleting the contents of '/home/runner/work/harmony-regression-tests/harmony-regression-tests'
Initializing the repository
Disabling automatic garbage collection
Setting up auth
***** /usr/bin/git lfs install --local
Updated Git hooks.
**** Git LFS initialized.
Fetching the repository
Determining the checkout info
***** Fetching LFS objects
Checking out the ref
/usr/bin/git log -1 --format='%H'
```


You should also be able to see the --build_arg shared_util=true for the nsidc-icesat2 docker build, but not any of the other images. they should be false.

flamingbear commented:
Lastly.  I did fix my fork build and you can examine the images and actions https://github.com/flamingbear/harmony-regression-tests/actions/runs/11034514374
I build HOSS and NSIDC-ICESAT2.

You can run the images by setting up your UAT env and then override the default images:
``` shell
export HARMONY_HOST_URL=https://harmony.uat.earthdata.nasa.gov
export AWS_ACCESS_KEY_ID=
export AWS_SECRET_ACCESS_KEY=
export REGRESSION_TESTS_NSIDC_ICESAT2_IMAGE=ghcr.io/flamingbear/regression-tests-nsidc-icesat2:latest
export REGRESSION_TESTS_HOSS_IMAGE=ghcr.io/flamingbear/regression-tests-hoss:latest
./run_notebooks.sh hoss nsidc-icesat2
```

You should see them running with the new images

``` shell
 ./run_notebooks.sh hoss nsidc-icesat2
Running regression tests
Test suite hoss starting
running test with ghcr.io/flamingbear/regression-tests-hoss:latest
Test suite nsidc-icesat2 starting
running test with ghcr.io/flamingbear/regression-tests-nsidc-icesat2:latest
```


They should run and pass against UAT.

flamingbear commented on README.md:
````diff
@@ -90,12 +90,15 @@ environment before installing from the environment.yml.
+       docker build -t ghcr.io/nasa/regression-tests-<new-suite-name>:latest -f ./Dockerfile --build-arg notebook=<new-test-notebook-name> --build-arg sub_dir=<new-suite-subdirectory> [--build-arg shared_utils=true] .
    ```
+
+1. If you would like to use shared utilities to help ease the coding you can include a build arg in
````
In what you ask? I don't know I'll go look

flamingbear replied:
3c6cea5

owenlittlejohns reviewed (approved):
I think the git-lfs changes all make sense. I re-cloned the repo (which was a bit slow, because the commits haven't been squashed yet), and the tests continue to run after rebuilding the image (although, they did take 37 minutes, which I'm a bit wary about - maybe I hit UAT at a bad time?).

The example runs of the workflow in your fork also look good.

Thanks for persisting with this!

flamingbear commented:
> (although, they did take 37 minutes, which I'm a bit wary about - maybe I hit UAT at a bad time?).

I think the nsidc-icesat2 tests are a bit slow, and since nothing is running in parallel it just adds onto the total time.

flamingbear commented:
@owenlittlejohns Yup, I checked and nsidc's runs in about 10min.
```
error	26-Sep-2024 09:09:17	
error	26-Sep-2024 09:09:17	Executing:   0%|          | 0/19 [00:00<?, ?cell/s]Executing notebook with kernel: python3
error	26-Sep-2024 09:09:17	
error	26-Sep-2024 09:09:17	Executing:   5%|▌         | 1/19 [00:09<02:46,  9.24s/cell]
error	26-Sep-2024 09:09:17	Executing:  26%|██▋       | 5/19 [00:14<00:34,  2.44s/cell]
error	26-Sep-2024 09:09:17	Executing:  37%|███▋      | 7/19 [00:19<00:30,  2.51s/cell]
error	26-Sep-2024 09:09:17	Executing:  47%|████▋     | 9/19 [00:20<00:18,  1.85s/cell]
error	26-Sep-2024 09:09:17	Executing:  63%|██████▎   | 12/19 [05:48<05:35, 47.96s/cell]
error	26-Sep-2024 09:09:17	Executing:  79%|███████▉  | 15/19 [06:49<02:29, 37.47s/cell]
error	26-Sep-2024 09:09:17	Executing:  95%|█████████▍| 18/19 [09:35<00:43, 43.69s/cell]
error	26-Sep-2024 09:09:17	Executing: 100%|██████████| 19/19 [09:38<00:00, 30.43s/cell]
build	26-Sep-2024 09:09:17	[0;32mTest suite nsidc-icesat2 succeeded[0m
```
